reactos/drivers/usb/usbhub_new/ioctl.c

```diff
@@ -0,0 +1,1424 @@
+
+    if (ConnectionName->ConnectionIndex == 0 ||
+        ConnectionName->ConnectionIndex > HubExtension->HubDescriptor->bNumberOfPorts)
+    {
+        Status = STATUS_INVALID_PARAMETER;
+        USBH_CompleteIrp(Irp, Status);
+        return Status;
+    }
+
+    PortDevice = HubExtension->PortData[ConnectionName->ConnectionIndex - 1].DeviceObject;
+
+    if (!PortDevice)
+    {
+        Irp->IoStatus.Information = sizeof(USB_NODE_CONNECTION_NAME);
+        ConnectionName->NodeName[0] = 0;
+        ConnectionName->ActualLength = sizeof(USB_NODE_CONNECTION_NAME);
+        USBH_CompleteIrp(Irp, Status);
+        return Status;
+    }
+
+    PortExtension = PortDevice->DeviceExtension;
+
+    if (!(PortExtension->PortPdoFlags & USBHUB_PDO_FLAG_HUB_DEVICE) ||
+        !(PortExtension->PortPdoFlags & USBHUB_PDO_FLAG_DEVICE_STARTED) ||
+        !(PortExtension->PortPdoFlags & USBHUB_PDO_FLAG_REG_DEV_INTERFACE))
+    {
+        ConnectionName->NodeName[0] = 0;
+        ConnectionName->ActualLength = sizeof(USB_NODE_CONNECTION_NAME);
+        Irp->IoStatus.Information = sizeof(USB_NODE_CONNECTION_NAME);
+        USBH_CompleteIrp(Irp, Status);
+        return Status;
+    }
+
+    LengthSkip = 0;
+
+    ASSERT(UNICODE_NULL == PortExtension->SymbolicLinkName.
+           Buffer[PortExtension->SymbolicLinkName.Length / sizeof(WCHAR)]);
+
+    if (PortExtension->SymbolicLinkName.Buffer[0] == L'\\')
+    {
+        BufferEnd = &PortExtension->SymbolicLinkName.Buffer[1];
+        Symbol = *BufferEnd;
+
+        if (PortExtension->SymbolicLinkName.Buffer[1] == L'\\')
+        {
+            LengthSkip = 2 * sizeof(WCHAR);
+        }
+        else
+        {
+            do
+            {
+                if (!Symbol)
+                {
+                    break;
+                }
+
+                BufferEnd += 1;
+                Symbol = *BufferEnd;
+            }
+            while (*BufferEnd != L'\\');
+
+            if (*BufferEnd == L'\\')
+            {
+                BufferEnd += 1;
+            }
+
+            LengthSkip = sizeof(WCHAR) *
+                         (BufferEnd - PortExtension->SymbolicLinkName.Buffer);
+        }
+    }
+
+    LengthName = PortExtension->SymbolicLinkName.Length - LengthSkip;
+
+    ConnectionName->ActualLength = 0;
+
+    RtlZeroMemory(ConnectionName->NodeName,
+                  Length - FIELD_OFFSET(USB_NODE_CONNECTION_NAME, NodeName));
+
+    LengthReturned = sizeof(USB_NODE_CONNECTION_NAME) + LengthName;
+
+    if (Length < LengthReturned)
+    {
+        ConnectionName->NodeName[0] = 0;
+        ConnectionName->ActualLength = LengthReturned;
+        Irp->IoStatus.Information = sizeof(USB_NODE_CONNECTION_NAME);
+        USBH_CompleteIrp(Irp, Status);
+        return Status;
+    }
+
+    RtlCopyMemory(&ConnectionName->NodeName[0],
+                  &PortExtension->SymbolicLinkName.Buffer[LengthSkip / sizeof(WCHAR)],
+                  LengthName);
+
+    ConnectionName->ActualLength = LengthReturned;
```

Comment:
I understand that. But `LengthReturned` is the size of the whole structure, whereas `ActualLength` should store the size of the `NodeName` string only, right?